tests/unittests/algo/test_space.py

```diff
@@ -1,7 +1,412 @@
+            for _ in range(8):
+                samples = dim.sample(8)
+                for sample in samples:
+                    assert sample in dim
+        assert "Improbable bounds" in str(exc.value)
+
+    def test_bad_bounds(self):
+        """Try setting bound with high <= low."""
+        with pytest.raises(ValueError):
+            Real('yolo', 'norm', 0, 2, low=+2, high=-2, shape=(4, 4))
+        with pytest.raises(ValueError):
+            Real('yolo', 'norm', 0, 2, low=+2, high=+2, shape=(4, 4))
+
+
+class TestInteger(object):
+    """Test methods of a `Integer` object."""
+
+    def test_as_you_are_as_you_were(self, seed):
+        """Test Integer.__init__."""
+        dim = Integer('yolo', 'uniform', -3, 6)
+        samples = dim.sample(seed=seed)
+        assert len(samples) == 1
+        assert samples[0] == -2
+
+        assert dists.uniform.interval(1.0, -3, 6) == dim.interval()
+        assert dists.uniform.interval(0.5, -3, 6) == dim.interval(0.5)
+
+        assert 1.0 in dim
+
+        assert str(dim) == "Integer(name=yolo, prior={uniform: (-3, 6), {}}, shape=())"
+
+        assert dim.name == 'yolo'
+        assert dim.type == 'integer'
+        assert dim.shape == ()
+
+    def test_contains(self):
+        """Check for integer test."""
+        dim = Integer('yolo', 'uniform', -3, 6)
+
+        assert 0.1 not in dim
+        assert (0.1, -0.2) not in dim
+        assert 0 in dim
+        assert (1, 2) not in dim
+        assert 6 not in dim
+        assert -3 in dim
+        assert -4 not in dim
+
+
+class TestCategorical(object):
+    """Test methods of a `Categorical` object."""
+
+    def test_with_tuple(self, seed):
+        """Test Categorical.__init__ with a tuple."""
+        categories = ('asdfa', 2)
+        dim = Categorical('yolo', categories)
+        samples = dim.sample(seed=seed)
+        assert len(samples) == 1
+        assert samples[0] == 'asdfa'
+        assert dim._probs == (0.5, 0.5)
+
+        assert categories == dim.interval()
+        assert categories == dim.interval(0.5)
+
+        assert 2 in dim
+        assert 3 not in dim
+
+        assert str(dim) == "Categorical(name=yolo, prior=[('asdfa', 0.5), (2, 0.5)], shape=())"
+
+        assert dim.name == 'yolo'
+        assert dim.type == 'categorical'
+        assert dim.shape == ()
+
+    def test_with_dict(self, seed):
+        """Test Categorical.__init__ with a dictionary."""
+        probs = (0.1, 0.2, 0.3, 0.4)
+        categories = ('asdfa', 2, 3, 4)
+        dim = Categorical('yolo', OrderedDict(zip(categories, probs)))
+        samples = dim.sample(seed=seed)
+        assert len(samples) == 1
+        assert samples[0] == 2
+        assert dim._probs == probs
+
+        assert categories == dim.interval()
+        assert categories == dim.interval(0.5)
+
+        assert 2 in dim
+        assert 0 not in dim
+
+        assert dim.name == 'yolo'
+        assert dim.type == 'categorical'
+        assert dim.shape == ()
```

Comment:
Should make a test with many sampling to ensure samples are close enough to probabilities, i.e., the proper probabilities have been used to sample categories.